File: niko-studio/src/store/openkl_contract.py
"""
OpenKL File System Contract

Implements the OpenKL "File as Truth" philosophy with standardized directory structure.

Directory Structure:
.writing/
├── store/
│   ├── sources/          # Original files
│   └── normalized/       # Normalized text (*.ok.md)
├── memories/
│   ├── by_date/          # Temporal organization (YYYY-MM/DD/<id>.md)
│   └── topics/           # Topic symlinks
├── citations/            # Citation JSON files
├── sessions/
│   ├── active/           # Active sessions
│   └── archived/         # Archived sessions
└── .ok/
    ├── kuzu/             # Kuzu graph database
    └── mapping.jsonl     # docID -> path mapping
"""

import json
import logging
import hashlib
import shutil
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Any, Iterator

logger = logging.getLogger("OpenKLContract")


@dataclass
class OpenKLPaths:
    """OpenKL directory paths configuration"""
    base_path: Path

    @property
    def store(self) -> Path:
        return self.base_path / "store"

    @property
    def sources(self) -> Path:
        return self.store / "sources"

    @property
    def normalized(self) -> Path:
        return self.store / "normalized"

    @property
    def memories(self) -> Path:
        return self.base_path / "memories"

    @property
    def memories_by_date(self) -> Path:
        return self.memories / "by_date"

    @property
    def memories_topics(self) -> Path:
        return self.memories / "topics"

    @property
    def citations(self) -> Path:
        return self.base_path / "citations"

    @property
    def sessions(self) -> Path:
        return self.base_path / "sessions"

    @property
    def active_sessions(self) -> Path:
        return self.sessions / "active"

    @property
    def archived_sessions(self) -> Path:
        return self.sessions / "archived"

    @property
    def ok_dir(self) -> Path:
        return self.base_path / ".ok"

    @property
    def kuzu_dir(self) -> Path:
        return self.ok_dir / "kuzu"

    @property
    def mapping_file(self) -> Path:
        return self.ok_dir / "mapping.jsonl"

    def all_paths(self) -> List[Path]:
        """Get all directory paths that should exist"""
        return [
            self.sources,
            self.normalized,
            self.memories_by_date,
            self.memories_topics,
            self.citations,
            self.active_sessions,
            self.archived_sessions,
            self.kuzu_dir,
        ]


@dataclass
class DocumentMapping:
    """Document ID to path mapping entry"""
    doc_id: str
    path: str
    sha256: str
    source_type: str
    created_at: str
    updated_at: str
    metadata: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "doc_id": self.doc_id,
            "path": self.path,
            "sha256": self.sha256,
            "source_type": self.source_type,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "DocumentMapping":
        return cls(
            doc_id=data["doc_id"],
            path=data["path"],
            sha256=data["sha256"],
            source_type=data.get("source_type", "general"),
            created_at=data["created_at"],
            updated_at=data["updated_at"],
            metadata=data.get("metadata", {}),
        )


class OpenKLContract:
    """
    OpenKL File System Contract Manager

    Implements the "File as Truth" philosophy:
    - All data is stored as files in a standardized directory structure
    - Documents are normalized and indexed
    - Mappings track document IDs to file paths
    - Supports temporal organization and topic linking
    """

    SUPPORTED_EXTENSIONS = {".md", ".txt", ".json", ".yaml", ".yml"}

    def __init__(self, base_path: str | Path = ".writing"):
        self.paths = OpenKLPaths(base_path=Path(base_path))
        self._mappings: Dict[str, DocumentMapping] = {}
        self._ensure_structure()
        self._load_mappings()

        logger.info(f"OpenKL contract initialized at: {self.paths.base_path}")

    def _ensure_structure(self) -> None:
        """Create all required directories"""
        for path in self.paths.all_paths():
            path.mkdir(parents=True, exist_ok=True)

        # Ensure mapping file exists
        if not self.paths.mapping_file.exists():
            self.paths.mapping_file.touch()

    def _load_mappings(self) -> None:
        """Load document mappings from JSONL file"""
        if not self.paths.mapping_file.exists():
            return

        try:
            with open(self.paths.mapping_file, "r", encoding="utf-8") as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        data = json.loads(line)
                        mapping = DocumentMapping.from_dict(data)
                        self._mappings[mapping.doc_id] = mapping
                    except json.JSONDecodeError as e:
                        logger.warning(f"Invalid mapping line: {e}")
        except Exception as e:
            logger.error(f"Failed to load mappings: {e}")

    def _save_mappings(self) -> None:
        """Save all mappings to JSONL file"""
        try:
            with open(self.paths.mapping_file, "w", encoding="utf-8") as f:
                for mapping in self._mappings.values():
                    f.write(json.dumps(mapping.to_dict(), ensure_ascii=False) + "\n")
        except Exception as e:
            logger.error(f"Failed to save mappings: {e}")

    def _append_mapping(self, mapping: DocumentMapping) -> None:
        """Append a single mapping to the JSONL file"""
        try:
            with open(self.paths.mapping_file, "a", encoding="utf-8") as f:
                f.write(json.dumps(mapping.to_dict(), ensure_ascii=False) + "\n")
        except Exception as e:
            logger.error(f"Failed to append mapping: {e}")

    @staticmethod
    def compute_sha256(content: str | bytes) -> str:
        """Compute SHA256 hash of content"""
        if isinstance(content, str):
            content = content.encode("utf-8")
        return hashlib.sha256(content).hexdigest()

    @staticmethod
    def generate_doc_id(path: Path, prefix: str = "doc") -> str:
        """Generate a unique document ID from path"""
        timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
        path_hash = hashlib.md5(str(path).encode()).hexdigest()[:8]
        return f"{prefix}-{timestamp}-{path_hash}"

    def normalize_content(self, content: str, source_path: Path) -> str:
        """
        Normalize document content to OpenKL format (*.ok.md)

        Normalization includes:
        - Unicode normalization
        - Line ending standardization
        - Metadata header addition
        """
        import unicodedata

        # Unicode NFC normalization
        content = unicodedata.normalize("NFC", content)

        # Standardize line endings
        content = content.replace("\r\n", "\n").replace("\r", "\n")

        # Add OpenKL metadata header
        now = datetime.now().isoformat()
        header = f"""---
source: {source_path.name}
normalized_at: {now}
format: ok.md
---

"""
        return header + content

    def ingest_file(
        self,
        source_path: Path,
        doc_id: Optional[str] = None,
        source_type: str = "general",
        normalize: bool = True,
        metadata: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Ingest a file into the OpenKL store

        Args:
            source_path: Path to the source file
            doc_id: Optional document ID (auto-generated if not provided)
            source_type: Document type classification
            normalize: Whether to create normalized version
            metadata: Additional metadata

        Returns:
            Document ID
        """
        source_path = Path(source_path)

        if not source_path.exists():
            raise FileNotFoundError(f"Source file not found: {source_path}")

        if source_path.suffix.lower() not in self.SUPPORTED_EXTENSIONS:
            raise ValueError(f"Unsupported file type: {source_path.suffix}")

        # Read content
        content = source_path.read_text(encoding="utf-8")
        sha256 = self.compute_sha256(content)

        # Generate doc_id if not provided
        if doc_id is None:
            doc_id = self.generate_doc_id(source_path)

        # Copy to sources directory
        dest_source = self.paths.sources / f"{doc_id}{source_path.suffix}"
        shutil.copy2(source_path, dest_source)

        # Create normalized version if requested
        if normalize:
            normalized_content = self.normalize_content(content, source_path)
            dest_normalized = self.paths.normalized / f"{doc_id}.ok.md"
            dest_normalized.write_text(normalized_content, encoding="utf-8")

        # Create mapping
        now = datetime.now().isoformat()
        mapping = DocumentMapping(
            doc_id=doc_id,
            path=str(dest_source.relative_to(self.paths.base_path)),
            sha256=sha256,
            source_type=source_type,
            created_at=now,
            updated_at=now,
            metadata=metadata or {},
        )

        self._mappings[doc_id] = mapping
        self._append_mapping(mapping)

        logger.info(f"Ingested document: {doc_id} from {source_path}")
        return doc_id

    def ingest_content(
        self,
        content: str,
        doc_id: Optional[str] = None,
        source_type: str = "general",
        filename: Optional[str] = None,
        normalize: bool = True,
        metadata: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Ingest content directly into the OpenKL store

        Args:
            content: Text content to ingest
            doc_id: Optional document ID
            source_type: Document type classification
            filename: Optional filename for the content
            normalize: Whether to create normalized version
            metadata: Additional metadata

        Returns:
            Document ID
        """
        sha256 = self.compute_sha256(content)

        # Generate doc_id if not provided
        if doc_id is None:
            timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
            hash_suffix = sha256[:8]
            doc_id = f"doc-{timestamp}-{hash_suffix}"

        # Determine filename
        if filename is None:
            filename = f"{doc_id}.md"

        # Save to sources
        dest_source = self.paths.sources / filename
        dest_source.write_text(content, encoding="utf-8")

        # Create normalized version if requested
        if normalize:
            normalized_content = self.normalize_content(content, dest_source)
            dest_normalized = self.paths.normalized / f"{doc_id}.ok.md"
            dest_normalized.write_text(normalized_content, encoding="utf-8")

        # Create mapping
        now = datetime.now().isoformat()
        mapping = DocumentMapping(
            doc_id=doc_id,
            path=str(dest_source.relative_to(self.paths.base_path)),
            sha256=sha256,
            source_type=source_type,
            created_at=now,
            updated_at=now,
            metadata=metadata or {},
        )

        self._mappings[doc_id] = mapping
        self._append_mapping(mapping)

        logger.info(f"Ingested content as document: {doc_id}")
        return doc_id

    def get_document(self, doc_id: str) -> Optional[Dict[str, Any]]:
        """Get document by ID"""
        mapping = self._mappings.get(doc_id)
        if not mapping:
            return None

        doc_path = self.paths.base_path / mapping.path
        if not doc_path.exists():
            logger.warning(f"Document file missing: {doc_path}")
            return None

        content = doc_path.read_text(encoding="utf-8")

        return {
            "doc_id": mapping.doc_id,
            "path": mapping.path,
            "content": content,
            "sha256": mapping.sha256,
            "source_type": mapping.source_type,
            "created_at": mapping.created_at,
            "updated_at": mapping.updated_at,
            "metadata": mapping.metadata,
        }

    def get_normalized_content(self, doc_id: str) -> Optional[str]:
        """Get normalized content for a document"""
        normalized_path = self.paths.normalized / f"{doc_id}.ok.md"
        if normalized_path.exists():
            return normalized_path.read_text(encoding="utf-8")
        return None

    def delete_document(self, doc_id: str) -> bool:
        """Delete a document and its normalized version"""
        mapping = self._mappings.get(doc_id)
        if not mapping:
            return False

        # Delete source file
        source_path = self.paths.base_path / mapping.path
        if source_path.exists():
            source_path.unlink()

        # Delete normalized file
        normalized_path = self.paths.normalized / f"{doc_id}.ok.md"
        if normalized_path.exists():
            normalized_path.unlink()

        # Remove mapping
        del self._mappings[doc_id]
        self._save_mappings()

        logger.info(f"Deleted document: {doc_id}")
        return True

    def list_documents(
        self,
        source_type: Optional[str] = None,
        limit: int = 100
    ) -> List[Dict[str, Any]]:
        """List all documents, optionally filtered by type"""
        results = []

        for mapping in self._mappings.values():
            if source_type and mapping.source_type != source_type:
                continue

            results.append({
                "doc_id": mapping.doc_id,
                "path": mapping.path,
                "source_type": mapping.source_type,
                "created_at": mapping.created_at,
                "metadata": mapping.metadata,
            })

            if len(results) >= limit:
                break

        return results

    def iter_documents(self, source_type: Optional[str] = None) -> Iterator[Dict[str, Any]]:
        """Iterate over all documents"""
        for mapping in self._mappings.values():
            if source_type and mapping.source_type != source_type:
                continue

            doc = self.get_document(mapping.doc_id)
            if doc:
                yield doc

    def create_memory(
        self,
        memory_id: str,
        content: str,
        topics: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Path:
        """
        Create a memory file in the temporal structure

        Args:
            memory_id: Memory ID
            content: Memory content
            topics: Optional list of topics for symlink creation
            metadata: Additional metadata

        Returns:
            Path to the created memory file
        """
        now = datetime.now()

        # Create date-based path
        date_dir = self.paths.memories_by_date / now.strftime("%Y-%m") / now.strftime("%d")
        date_dir.mkdir(parents=True, exist_ok=True)

        # Write memory file
        memory_path = date_dir / f"{memory_id}.md"

        # Add metadata header
        header = f"""---
memory_id: {memory_id}
created_at: {now.isoformat()}
topics: {topics or []}
metadata: {json.dumps(metadata or {})}
---

"""
        memory_path.write_text(header + content, encoding="utf-8")

        # Create topic symlinks
        if topics:
            for topic in topics:
                topic_dir = self.paths.memories_topics / topic
                topic_dir.mkdir(parents=True, exist_ok=True)

                symlink_path = topic_dir / f"{memory_id}.md"
                if not symlink_path.exists():
                    try:
                        symlink_path.symlink_to(memory_path)
                    except OSError:
                        # Symlinks may not work on all systems
                        pass

        logger.info(f"Created memory: {memory_id}")
        return memory_path

    def save_citation(
        self,
        citation_id: str,
        citation_data: Dict[str, Any]
    ) -> Path:
        """Save a citation to the citations directory"""
        citation_path = self.paths.citations / f"{citation_id}.json"
        citation_path.write_text(
            json.dumps(citation_data, ensure_ascii=False, indent=2),
            encoding="utf-8"
        )
        return citation_path

    def get_citation(self, citation_id: str) -> Optional[Dict[str, Any]]:
        """Get a citation by ID"""
        citation_path = self.paths.citations / f"{citation_id}.json"
        if not citation_path.exists():
            return None

        return json.loads(citation_path.read_text(encoding="utf-8"))

    def verify_integrity(self) -> Dict[str, Any]:
        """Verify the integrity of the store"""
        results = {
            "total_documents": len(self._mappings),
            "missing_files": [],
            "hash_mismatches": [],
            "orphaned_files": [],
        }

        # Check mapped documents
        for doc_id, mapping in self._mappings.items():
            doc_path = self.paths.base_path / mapping.path

            if not doc_path.exists():
                results["missing_files"].append(doc_id)
                continue

            # Verify hash
            content = doc_path.read_text(encoding="utf-8")
            current_hash = self.compute_sha256(content)

            if current_hash != mapping.sha256:
                results["hash_mismatches"].append({
                    "doc_id": doc_id,
                    "expected": mapping.sha256,
                    "actual": current_hash,
                })

        # Check for orphaned files
        mapped_paths = {m.path for m in self._mappings.values()}

        for source_file in self.paths.sources.glob("*"):
            rel_path = str(source_file.relative_to(self.paths.base_path))
            if rel_path not in mapped_paths:
                results["orphaned_files"].append(rel_path)

        return results

File: src-ts/store/openkl-contract.ts
/**
 * OpenKL File System Contract
 *
 * Implements the OpenKL "File as Truth" philosophy with standardized directory structure.
 *
 * Directory Structure:
 * .writing/
 *   store/
 *     sources/          # Original files
 *     normalized/       # Normalized text (*.ok.md)
 *   memories/
 *     by_date/          # Temporal organization (YYYY-MM/DD/<id>.md)
 *     topics/           # Topic symlinks
 *   citations/            # Citation JSON files
 *   sessions/
 *     active/           # Active sessions
 *     archived/         # Archived sessions
 *   .ok/
 *     kuzu/             # Kuzu graph database
 *     mapping.jsonl     # docID -> path mapping
 */

import { randomUUID } from 'node:crypto';
import {
  writeFileSync,
  readFileSync,
  appendFileSync,
  existsSync,
  mkdirSync,
  unlinkSync,
  copyFileSync,
  readdirSync,
  symlinkSync,
} from 'node:fs';
import { join, dirname, basename, extname, relative } from 'node:path';
import { homedir, platform } from 'node:os';
import { createHash } from 'node:crypto';

// ---------------------------------------------------------------------------
// OpenKLPaths
// ---------------------------------------------------------------------------

/**
 * OpenKL directory paths configuration
 */
export class OpenKLPaths {
  basePath: string;

  constructor(basePath: string = '.writing') {
    this.basePath = basePath;
  }

  get store(): string {
    return join(this.basePath, 'store');
  }

  get sources(): string {
    return join(this.store, 'sources');
  }

  get normalized(): string {
    return join(this.store, 'normalized');
  }

  get memories(): string {
    return join(this.basePath, 'memories');
  }

  get memoriesByDate(): string {
    return join(this.memories, 'by_date');
  }

  get memoriesTopics(): string {
    return join(this.memories, 'topics');
  }

  get citations(): string {
    return join(this.basePath, 'citations');
  }

  get sessions(): string {
    return join(this.basePath, 'sessions');
  }

  get activeSessions(): string {
    return join(this.sessions, 'active');
  }

  get archivedSessions(): string {
    return join(this.sessions, 'archived');
  }

  get okDir(): string {
    return join(this.basePath, '.ok');
  }

  get kuzuDir(): string {
    return join(this.okDir, 'kuzu');
  }

  get mappingFile(): string {
    return join(this.okDir, 'mapping.jsonl');
  }

  /** Get all directory paths that should exist */
  allPaths(): string[] {
    return [
      this.sources,
      this.normalized,
      this.memoriesByDate,
      this.memoriesTopics,
      this.citations,
      this.activeSessions,
      this.archivedSessions,
      this.kuzuDir,
    ];
  }
}

// ---------------------------------------------------------------------------
// DocumentMapping
// ---------------------------------------------------------------------------

/**
 * Document ID to path mapping entry
 */
export class DocumentMapping {
  docId: string;
  path: string;
  sha256: string;
  sourceType: string;
  createdAt: string;
  updatedAt: string;
  metadata: Record<string, unknown>;

  constructor(params: {
    docId: string;
    path: string;
    sha256: string;
    sourceType: string;
    createdAt: string;
    updatedAt: string;
    metadata?: Record<string, unknown>;
  }) {
    this.docId = params.docId;
    this.path = params.path;
    this.sha256 = params.sha256;
    this.sourceType = params.sourceType;
    this.createdAt = params.createdAt;
    this.updatedAt = params.updatedAt;
    this.metadata = params.metadata ?? {};
  }

  toDict(): Record<string, unknown> {
    return {
      doc_id: this.docId,
      path: this.path,
      sha256: this.sha256,
      source_type: this.sourceType,
      created_at: this.createdAt,
      updated_at: this.updatedAt,
      metadata: this.metadata,
    };
  }

  static fromDict(data: Record<string, unknown>): DocumentMapping {
    return new DocumentMapping({
      docId: data['doc_id'] as string,
      path: data['path'] as string,
      sha256: data['sha256'] as string,
      sourceType: (data['source_type'] as string) ?? 'general',
      createdAt: data['created_at'] as string,
      updatedAt: data['updated_at'] as string,
      metadata: (data['metadata'] as Record<string, unknown>) ?? {},
    });
  }
}

// ---------------------------------------------------------------------------
// OpenKLContract
// ---------------------------------------------------------------------------

/**
 * OpenKL File System Contract Manager
 *
 * Implements the "File as Truth" philosophy:
 * - All data is stored as files in a standardized directory structure
 * - Documents are normalized and indexed
 * - Mappings track document IDs to file paths
 * - Supports temporal organization and topic linking
 */
export class OpenKLContract {
  static readonly SUPPORTED_EXTENSIONS: readonly string[] = ['.md', '.txt', '.json', '.yaml', '.yml'];

  paths: OpenKLPaths;
  private _mappings: Map<string, DocumentMapping> = new Map();

  constructor(basePath: string = '.writing') {
    this.paths = new OpenKLPaths(basePath);
    this._ensureStructure();
    this._loadMappings();

    console.info(`OpenKL contract initialized at: ${this.paths.basePath}`);
  }

  // -----------------------------------------------------------------------
  // Directory management
  // -----------------------------------------------------------------------

  private _ensureStructure(): void {
    /** Create all required directories */
    for (const path of this.paths.allPaths()) {
      mkdirSync(path, { recursive: true });
    }
    // Ensure mapping file exists
    if (!existsSync(this.paths.mappingFile)) {
      writeFileSync(this.paths.mappingFile, '');
    }
  }

  // -----------------------------------------------------------------------
  // Mapping persistence
  // -----------------------------------------------------------------------

  private _loadMappings(): void {
    /** Load document mappings from JSONL file */
    if (!existsSync(this.paths.mappingFile)) return;
    try {
      const content = readFileSync(this.paths.mappingFile, 'utf-8');
      for (const line of content.split('\n')) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        try {
          const data = JSON.parse(trimmed);
          const mapping = DocumentMapping.fromDict(data);
          this._mappings.set(mapping.docId, mapping);
        } catch (e) {
          if (e instanceof SyntaxError) {
            console.warn(`Invalid mapping line: ${e.message}`);
          }
        }
      }
    } catch (e) {
      console.error(`Failed to load mappings: ${e}`);
    }
  }

  private _saveMappings(): void {
    /** Save all mappings to JSONL file */
    try {
      const lines = Array.from(this._mappings.values()).map((m) =>
        JSON.stringify(m.toDict()) + '\n'
      );
      writeFileSync(this.paths.mappingFile, lines.join(''), 'utf-8');
    } catch (e) {
      console.error(`Failed to save mappings: ${e}`);
    }
  }

  private _appendMapping(mapping: DocumentMapping): void {
    /** Append a single mapping to the JSONL file */
    try {
      const line = JSON.stringify(mapping.toDict()) + '\n';
      appendFileSync(this.paths.mappingFile, line, 'utf-8');
    } catch (e) {
      console.error(`Failed to append mapping: ${e}`);
    }
  }

  // -----------------------------------------------------------------------
  // Static helpers
  // -----------------------------------------------------------------------

  /** Compute SHA256 hash of content */
  static computeSha256(content: string | Buffer): string {
    if (typeof content === 'string') {
      content = Buffer.from(content, 'utf-8');
    }
    return createHash('sha256').update(content).digest('hex');
  }

  /** Generate a unique document ID from path */
  static generateDocId(path: string, prefix: string = 'doc'): string {
    const timestamp = new Date().toISOString().replace(/[-:T]/g, '').replace(/\./g, '');
    const pathHash = createHash('md5').update(path).digest('hex').slice(0, 8);
    return `${prefix}-${timestamp}-${pathHash}`;
  }

  /**
   * Normalize document content to OpenKL format (*.ok.md)
   *
   * Normalization includes:
   * - Unicode NFC normalization
   * - Line ending standardization
   * - Metadata header addition
   */
  normalizeContent(content: string, sourcePath: string): string {
    // Unicode NFC normalization (JavaScript strings are already NFC)
    // Standardize line endings
    content = content.replace(/\r\n/g, '\n').replace(/\r/g, '\n');

    // Add OpenKL metadata header
    const now = new Date().toISOString();
    const header = `---
source: ${basename(sourcePath)}
normalized_at: ${now}
format: ok.md

`;

    return header + content;
  }

  // -----------------------------------------------------------------------
  // File ingestion
  // -----------------------------------------------------------------------

  /**
   * Ingest a file into the OpenKL store
   *
   * @param sourcePath - Path to the source file
   * @param docId - Optional document ID (auto-generated if not provided)
   * @param sourceType - Document type classification
   * @param normalize - Whether to create normalized version
   * @param metadata - Additional metadata
   * @returns Document ID
   */
  ingestFile(
    sourcePath: string,
    docId?: string | null,
    sourceType: string = 'general',
    normalize: boolean = true,
    metadata?: Record<string, unknown> | null
  ): string {
    if (!existsSync(sourcePath)) {
      throw new Error(`Source file not found: ${sourcePath}`);
    }

    const ext = extname(sourcePath).toLowerCase();
    if (!OpenKLContract.SUPPORTED_EXTENSIONS.includes(ext)) {
      throw new Error(`Unsupported file type: ${ext}`);
    }

    // Read content
    const content = readFileSync(sourcePath, 'utf-8');
    const sha256 = OpenKLContract.computeSha256(content);

    // Generate doc_id if not provided
    if (docId === undefined || docId === null) {
      docId = OpenKLContract.generateDocId(sourcePath);
    }

    // Copy to sources directory
    const destSource = join(this.paths.sources, `${docId}${extname(sourcePath)}`);
    copyFileSync(sourcePath, destSource);

    // Create normalized version if requested
    if (normalize) {
      const normalizedContent = this.normalizeContent(content, sourcePath);
      const destNormalized = join(this.paths.normalized, `${docId}.ok.md`);
      writeFileSync(destNormalized, normalizedContent, 'utf-8');
    }

    // Create mapping
    const now = new Date().toISOString();
    const mapping = new DocumentMapping({
      docId,
      path: relative(this.paths.basePath, destSource),
      sha256,
      sourceType,
      createdAt: now,
      updatedAt: now,
      metadata: metadata ?? {},
    });

    this._mappings.set(docId, mapping);
    this._appendMapping(mapping);

    console.info(`Ingested document: ${docId} from ${sourcePath}`);
    return docId;
  }

  /**
   * Ingest content directly into the OpenKL store
   *
   * @param content - Text content to ingest
   * @param docId - Optional document ID
   * @param sourceType - Document type classification
   * @param filename - Optional filename for the content
   * @param normalize - Whether to create normalized version
   * @param metadata - Additional metadata
   * @returns Document ID
   */
  ingestContent(
    content: string,
    docId?: string | null,
    sourceType: string = 'general',
    filename?: string | null,
    normalize: boolean = true,
    metadata?: Record<string, unknown> | null
  ): string {
    const sha256 = OpenKLContract.computeSha256(content);

    // Generate doc_id if not provided
    if (docId === undefined || docId === null) {
      const timestamp = new Date().toISOString().replace(/[-:T]/g, '').replace(/\./g, '');
      const hashSuffix = sha256.slice(0, 8);
      docId = `doc-${timestamp}-${hashSuffix}`;
    }

    // Determine filename
    if (filename === undefined || filename === null) {
      filename = `${docId}.md`;
    }

    // Save to sources
    const destSource = join(this.paths.sources, filename);
    writeFileSync(destSource, content, 'utf-8');

    // Create normalized version if requested
    if (normalize) {
      const normalizedContent = this.normalizeContent(content, destSource);
      const destNormalized = join(this.paths.normalized, `${docId}.ok.md`);
      writeFileSync(destNormalized, normalizedContent, 'utf-8');
    }

    // Create mapping
    const now = new Date().toISOString();
    const mapping = new DocumentMapping({
      docId,
      path: relative(this.paths.basePath, destSource),
      sha256,
      sourceType,
      createdAt: now,
      updatedAt: now,
      metadata: metadata ?? {},
    });

    this._mappings.set(docId, mapping);
    this._appendMapping(mapping);

    console.info(`Ingested content as document: ${docId}`);
    return docId;
  }

  // -----------------------------------------------------------------------
  // Document CRUD
  // -----------------------------------------------------------------------

  /** Get document by ID */
  getDocument(docId: string): Record<string, unknown> | null {
    const mapping = this._mappings.get(docId);
    if (!mapping) {
      return null;
    }

    const docPath = join(this.paths.basePath, mapping.path);
    if (!existsSync(docPath)) {
      console.warn(`Document file missing: ${docPath}`);
      return null;
    }

    const content = readFileSync(docPath, 'utf-8');
    return {
      doc_id: mapping.docId,
      path: mapping.path,
      content,
      sha256: mapping.sha256,
      source_type: mapping.sourceType,
      created_at: mapping.createdAt,
      updated_at: mapping.updatedAt,
      metadata: mapping.metadata,
    };
  }

  /** Get normalized content for a document */
  getNormalizedContent(docId: string): string | null {
    const normalizedPath = join(this.paths.normalized, `${docId}.ok.md`);
    if (existsSync(normalizedPath)) {
      return readFileSync(normalizedPath, 'utf-8');
    }
    return null;
  }

  /** Delete a document and its normalized version */
  deleteDocument(docId: string): boolean {
    const mapping = this._mappings.get(docId);
    if (!mapping) {
      return false;
    }

    // Delete source file
    const sourcePath = join(this.paths.basePath, mapping.path);
    if (existsSync(sourcePath)) {
      unlinkSync(sourcePath);
    }

    // Delete normalized file
    const normalizedPath = join(this.paths.normalized, `${docId}.ok.md`);
    if (existsSync(normalizedPath)) {
      unlinkSync(normalizedPath);
    }

    // Remove mapping
    this._mappings.delete(docId);
    this._saveMappings();

    console.info(`Deleted document: ${docId}`);
    return true;
  }

  /** List all documents, optionally filtered by type */
  listDocuments(sourceType?: string | null, limit: number = 100): Record<string, unknown>[] {
    const results: Record<string, unknown>[] = [];

    for (const mapping of this._mappings.values()) {
      if (sourceType && mapping.sourceType !== sourceType) {
        continue;
      }

      results.push({
        doc_id: mapping.docId,
        path: mapping.path,
        source_type: mapping.sourceType,
        created_at: mapping.createdAt,
        metadata: mapping.metadata,
      });

      if (results.length >= limit) {
        break;
      }
    }

    return results;
  }

  /** Iterate over all documents */
  *iterDocuments(sourceType?: string | null): Generator<Record<string, unknown>> {
    for (const mapping of this._mappings.values()) {
      if (sourceType && mapping.sourceType !== sourceType) {
        continue;
      }
      const doc = this.getDocument(mapping.docId);
      if (doc) {
        yield doc;
      }
    }
  }

  // -----------------------------------------------------------------------
  // Memory operations
  // -----------------------------------------------------------------------

  /**
   * Create a memory file in the temporal structure
   *
   * @param memoryId - Memory ID
   * @param content - Memory content
   * @param topics - Optional list of topics for symlink creation
   * @param metadata - Additional metadata
   * @returns Path to the created memory file
   */
  createMemory(
    memoryId: string,
    content: string,
    topics?: string[] | null,
    metadata?: Record<string, unknown> | null
  ): string {
    const now = new Date();

    // Create date-based path
    const yearMonth = now.toISOString().slice(0, 7);
    const day = now.toISOString().slice(8, 10);
    const dateDir = join(this.paths.memoriesByDate, yearMonth, day);
    mkdirSync(dateDir, { recursive: true });

    // Write memory file
    const memoryPath = join(dateDir, `${memoryId}.md`);

    // Add metadata header
    const header = `---
memory_id: ${memoryId}
created_at: ${now.toISOString()}
topics: ${JSON.stringify(topics ?? [])}
metadata: ${JSON.stringify(metadata ?? {})}

`;

    writeFileSync(memoryPath, header + content, 'utf-8');

    // Create topic symlinks
    if (topics) {
      for (const topic of topics) {
        const topicDir = join(this.paths.memoriesTopics, topic);
        mkdirSync(topicDir, { recursive: true });

        const symlinkPath = join(topicDir, `${memoryId}.md`);
        if (!existsSync(symlinkPath)) {
          try {
            if (platform() === 'win32') {
              // Windows: copy instead of symlink
              copyFileSync(memoryPath, symlinkPath);
            } else {
              symlinkSync(memoryPath, symlinkPath);
            }
          } catch {
            // Symlinks may not work on all systems
          }
        }
      }
    }

    console.info(`Created memory: ${memoryId}`);
    return memoryPath;
  }

  // -----------------------------------------------------------------------
  // Citation operations
  // -----------------------------------------------------------------------

  /** Save a citation to the citations directory */
  saveCitation(citationId: string, citationData: Record<string, unknown>): string {
    const citationPath = join(this.paths.citations, `${citationId}.json`);
    writeFileSync(
      citationPath,
      JSON.stringify(citationData, null, 2),
      'utf-8'
    );
    return citationPath;
  }

  /** Get a citation by ID */
  getCitation(citationId: string): Record<string, unknown> | null {
    const citationPath = join(this.paths.citations, `${citationId}.json`);
    if (!existsSync(citationPath)) {
      return null;
    }
    return JSON.parse(readFileSync(citationPath, 'utf-8'));
  }

  // -----------------------------------------------------------------------
  // Integrity verification
  // -----------------------------------------------------------------------

  /** Verify the integrity of the store */
  verifyIntegrity(): Record<string, unknown> {
    const results: Record<string, unknown> = {
      total_documents: this._mappings.size,
      missing_files: [] as string[],
      hash_mismatches: [] as Record<string, unknown>[],
      orphaned_files: [] as string[],
    };

    // Check mapped documents
    for (const [docId, mapping] of this._mappings) {
      const docPath = join(this.paths.basePath, mapping.path);

      if (!existsSync(docPath)) {
        (results.missing_files as string[]).push(docId);
        continue;
      }

      // Verify hash
      const content = readFileSync(docPath, 'utf-8');
      const currentHash = OpenKLContract.computeSha256(content);

      if (currentHash !== mapping.sha256) {
        (results.hash_mismatches as Record<string, unknown>[]).push({
          doc_id: docId,
          expected: mapping.sha256,
          actual: currentHash,
        });
      }
    }

    // Check for orphaned files
    const mappedPaths = new Set(
      Array.from(this._mappings.values()).map((m) => m.path)
    );

    try {
      const sourceFiles = readdirSync(this.paths.sources, { withFileTypes: true });
      for (const sourceFile of sourceFiles) {
        if (sourceFile.isFile()) {
          const relPath = relative(this.paths.basePath, join(this.paths.sources, sourceFile.name));
          if (!mappedPaths.has(relPath)) {
            (results.orphaned_files as string[]).push(relPath);
          }
        }
      }
    } catch {
      // Sources directory may not exist yet
    }

    return results;
  }
}
